dnf/transaction_sr.py

```diff
@@ -0,0 +1,508 @@
+
+                    return
+
+                self._raise_or_warn(self._ignore_installed,
+                    _('Rpm nevra "{nevra}" not installed for action "{action}".').format(nevra=nevra, action=action))
+
+                return
+
+            # erasing the original version (the reverse part of an action like
+            # e.g. upgrade) is more robust, but we can't do it if
+            # skip_unavailable is True, because if the forward part of the
+            # action is skipped, we would simply remove the package here
+            if not self._skip_unavailable or action == "Removed":
+                self._base.goal.erase(query[0], clean_deps=False)
+        elif action == "Reason Change":
+            self._base.history.set_reason(query[0], reason)
+        else:
+            raise TransactionError(
+                _('Unexpected value of package action "{action}" for rpm nevra "{nevra}".')
+                    .format(action=action, nevra=nevra)
+            )
+
+    def get_data(self):
+        """
+        :returns: the loaded data of the transaction
+        """
+
+        return self._replay_data
+
+    def get_warnings(self):
+        """
+        :returns: an array of warnings gathered during the transaction replay
+        """
+
+        return self._warnings
+
+    def run(self):
+        """
+        Replays the transaction.
+        """
+
+        # Override configuration options that affect how the transaction is resolved
+        # TODO if used through API, we shouldn't change the global state here?
+        self._base.conf.clean_requirements_on_remove = False
+        self._base.conf.install_weak_deps = False
+
+        fn = self._filename
+        errors = []
+
+        for pkg_data in self._rpms:
+            try:
+                self._replay_pkg_action(pkg_data)
+            except TransactionError as e:
+                errors.append(e)
+
+        for group_data in self._groups:
+            try:
+                action = group_data["action"]
+                group_id = group_data["id"]
+
+                if action == "Install":
+                    try:
+                        pkg_types = libdnf.transaction.stringToCompsPackageType(group_data["package_types"])
+                    except libdnf.error.Error as e:
+                        errors.append(TransactionError(str(e)))
+                        continue
+
+                    self._base.group_install(group_id, pkg_types)
+                elif action == "Upgrade":
+                    self._base.group_upgrade(group_id)
+                elif action == "Removed":
+                    self._base.group_remove(group_id)
+                else:
+                    errors.append(TransactionError(
+                        _('Unexpected value of group action "{action}" for group "{group}".')
+                            .format(action=action, group=group_id)
+                    ))
+            except KeyError as e:
+                errors.append(TransactionError(
```

Comment:
I don't think `opts._ignore_installed` is relevant here, but `opts._skip_unavailable` is, it seems that's what you meant? Yes, this switch remains to be implemented for groups/envs, there's the TODO just a few lines below the one you commented on. As it says, I can't distinguish the case since I always get the same `CompsError`, we need to discuss this.